### PR TITLE
Disable reloc shader when desc range value count != 0

### PR DESCRIPTION
**Pull request by s-perron**
An unlinked shader compilation where the user data nodes are not
available cannot tell if a descriptor for a sampler is actaully
an "immutable value", so the code will be generated the same way
it is for any other descriptor.  This means that pipelines with
immutable samplers cannot be compiled using relocatable shader.

### CODE REVIEW

**Comment by trenouf** [comment]
Doesn't the driver also set them up as normal samplers?

File: docs/DdnRelocatableShaderElf.md [hunk 1]
```diff
@@ -137,9 +137,9 @@ top are the three "top-level" nodes. They are considered the top because the can
 be accessed directly. On the other hand, to access the descriptor for say
 descriptor set 0 and binding 2, the DescriptorTableVaPtr must be loaded first.
 
-Buffer descriptors are placed in the top-level are called "dynamic
-descriptors", and they are used to implement the feature where the application
-can modify the start offset of buffer without rebuilding the pipeline state.
+Buffer descriptors are placed in the top-level are called "dynamic descriptors",
```
Comment (nhaehnle) [inline docs/DdnRelocatableShaderElf.md:140]: *that* are placed?

This change seems unrelated to what the commit message talks about.

File: docs/DdnRelocatableShaderElf.md [hunk 2]
```diff
@@ -183,14 +183,22 @@ offset of the descriptor table that contains descriptor set 0. This happens in
...
-driver will not use compact buffer descriptors.  Compact buffer descriptors
-can only ever be top-level nodes.
+While building an unlinked shader, it is not possible to know if compact buffer
+descriptors will be used or not. It is unclear if it would be worth generating
+code general enough to work in both cases. The extra overhead would potentially
+negate any gains from using compact buffer descriptors. Therefore unlinked
+shaders are compiled assuming that there are no compact buffer descriptors. If
+the XGL option `EnableRelocatableShaders` is true, then the driver will not use
+compact buffer descriptors. Compact buffer descriptors can only ever be
+top-level nodes.
```
Comment (nhaehnle) [inline docs/DdnRelocatableShaderElf.md:193]: Unrelated change?

**Comment by nhaehnle** [comment]
> Doesn't the driver also set them up as normal samplers?

This does not actually appear to be the case today. Maybe it should be...

File: docs/DdnRelocatableShaderElf.md (hunk 1 above)
Comment (s-perron) [reply docs/DdnRelocatableShaderElf.md:140]: I ran a markdown formatter on the document, and it changed some of the other code.  I can undo that.

File: docs/DdnRelocatableShaderElf.md (hunk 2 above)
Comment (s-perron) [reply docs/DdnRelocatableShaderElf.md:193]: Just formatting changes.

**Comment by s-perron** [comment]
> Doesn't the driver also set them up as normal samplers?

I've got traces that fail because this does not happen.  If we could get that done, we could remove this check.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f5fcfd69ada8a0e3ba1683dd7fdcc06879bf82f3
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev c9937b277c491a55c689eec7cdd48fb238b8004c</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev c6c90450fb1abf5f414acf1e38a0f51a72c426c1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev d3f03b36d914d6f4c4d27709bf29b89c4803c9cf</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/131664)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 24006/43895 (54.7%)</li>
<li>Failed: 0/43895 (0.0%)</li>
<li>Skipped: 19889/43895 (45.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 24005/43895 (54.7%)</li>
<li>Failed: 0/43895 (0.0%)</li>
<li>Skipped: 19890/43895 (45.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 22647/43874 (51.6%)</li>
<li>Failed: 0/43874 (0.0%)</li>
<li>Skipped: 21227/43874 (48.4%)</li>
</ul></details>


</ul></details>

**Review by trenouf — APPROVED** [review]
LGTM. I guess you want to do this now, as the proposed driver change won't happen immediately.

**Comment by s-perron** [comment]
> LGTM. I guess you want to do this now, as the proposed driver change won't happen immediately.

Yes, please and thank you.
